File: front/lib/domain/model/owner.dart
import 'package:amap_en_ligne/domain/model/user_preferences.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'owner.freezed.dart';
part 'owner.g.dart';

/// Account status for an instance-level owner.
enum AccountStatus {
  @JsonValue('ACTIVE')
  active,
  @JsonValue('SUSPENDED')
  suspended,
}

/// Instance-level owner. Presence of a row in the local cache materialises
/// the OWNER role. There is no [roles] field — the role is implicit.
///
/// Mirrors `persistence.model.Owner` on the back.
@freezed
abstract class Owner with _$Owner {
  const factory Owner({
    @JsonKey(name: 'owner_id') required String ownerId,
    @JsonKey(name: 'first_name') required String firstName,
    @JsonKey(name: 'last_name') required String lastName,
    required String email,
    String? phone,
    @JsonKey(name: 'account_status')
    @Default(AccountStatus.active)
    AccountStatus accountStatus,
    @JsonKey(name: 'registered_at') required String registeredAt,
    @JsonKey(name: 'updated_at') required String updatedAt,
    @JsonKey(name: 'user_preferences') UserPreferences? userPreferences,
  }) = _Owner;

  factory Owner.fromJson(Map<String, Object?> json) => _$OwnerFromJson(json);
}

File: back/persistence/model/src/main/kotlin/Owner.kt
@file:OptIn(ExperimentalTime::class)

package persistence.model

import id.Id
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime
import kotlin.time.Instant

@Serializable
enum class AccountStatus {
    ACTIVE,
    SUSPENDED,
}

/**
 * Instance-level owner. Presence of a row in this table materialises the OWNER role.
 * There is no [roles] field — the role is implicit.
 *
 * After the sub/id unification, [ownerId] equals the auth-provider subject (sub).
 * The [sub] field has been removed — callers that previously read [sub] should use
 * [ownerId] instead.
 */
@Serializable
data class Owner(
    @SerialName("owner_id") val ownerId: Id<Owner>,
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    val email: String,
    val phone: String? = null,
    @SerialName("account_status") val accountStatus: AccountStatus = AccountStatus.ACTIVE,
    @SerialName("registered_at") val registeredAt: Instant,
    @SerialName("updated_at") val updatedAt: Instant,
    @SerialName("user_preferences") val userPreferences: UserPreferences =
        UserPreferences(
            emailNotificationsEnabled = true,
            pushNotificationsEnabled = false,
            lastUpdatedInstant = Instant.fromEpochMilliseconds(0L),
        ),
)
